We rarely get our data in just the form we want.  In this worksheet we will investigate some techniques for making relatively simple adjustments that help us get _from our input data into a data frame that is ready for us to use_.

We have some data in the file `iris.csv` which has been hand-transcribed from notebooks and contains a few errors.  To clean the data we will:
  * Load up the data
  * Understand the Data
  * Remove empty values
  * Fix format errors
  * Fix incorrect data
  * Remove duplicates

After this, we are ready to visualise the data (which we will learn more about next week).

In [228]:
import pandas as pd

# load data - notice that iris data has no index, so we will use a fresh one
iris = pd.read_csv("data/iris.csv")
iris


sepal length  sepal width  petal length  petal width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
147           6.7          3.0           5.2          2.3  Iris-virginica
148           6.3          2.5           5.0          1.9  Iris-virginica
149           6.5          3.0           5.2          2.0  Iris-virginica
150           6.2          3.4           5.4          2.3  Iris-virginica
151           5.9          3.0           5.1          1.8  Iris-virginica

[152 rows x 5 columns]

Jupyter always gives us a preview of our data when we print it, but we can specifially ask for the `head` rows, the `tail` rows and `info` about any dataframe.  `info` in particular gives us very useful information about our data.  Pay close attention to the "non-null" count and the data type of each column.


In [229]:
print(iris.head(2))
print("")
print(iris.tail(2))
print("")
print(iris.info())

   sepal length  sepal width  petal length  petal width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa

     sepal length  sepal width  petal length  petal width           class
150           6.2          3.4           5.4          2.3  Iris-virginica
151           5.9          3.0           5.1          1.8  Iris-virginica

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 152 entries, 0 to 151
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal length  152 non-null    float64
 1   sepal width   151 non-null    float64
 2   petal length  152 non-null    float64
 3   petal width   152 non-null    float64
 4   class         152 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.1+ KB
None




The first issue that shows up is the ***missing value in "sepal width".*** Every other column has 152 values, but this one has just ***151.*** Missing values in a table are called "null" values and they can mess with our analysis if we are not aware of them. Sometimes we do want to leave them there, but often we want to exclude that data.

**YOU MUST NEVER MODIFY YOUR SOURCE DATA**

If there is "junk" in your input file, never make changes directly in the file, for the following reasons:
  * data is often audited and **modification of official data could be an infringement**
  * one person's junk is another person's treasure
  * pandas can adjust the data for you easily so you can have a "clean" version without adjusting the original

Lets start by finding that null value.  We know it is in the "sepal width" series



In [230]:
iris["sepal width"]

0      3.5
1      3.0
2      3.2
3      3.1
4      3.6
      ... 
147    3.0
148    2.5
149    3.0
150    3.4
151    3.0
Name: sepal width, Length: 152, dtype: float64

we can't see it in the preview, but pandas can filter a series to keep only the null values, but do that, we will need to understand indexing with masks.   

Firstly, we get the "sepal width" series from the data frame.  Then we call a function on that series which will convert all values to true or false.  If the original value was not null, we will get false, if it was null, we get true.  You will have to believe me, but there is a single "True" in there.

In [231]:
mask = iris["sepal width"].isnull()
print(mask.info())
mask

<class 'pandas.core.series.Series'>
RangeIndex: 152 entries, 0 to 151
Series name: sepal width
Non-Null Count  Dtype
--------------  -----
152 non-null    bool 
dtypes: bool(1)
memory usage: 284.0 bytes
None


0      False
1      False
2      False
3      False
4      False
       ...  
147    False
148    False
149    False
150    False
151    False
Name: sepal width, Length: 152, dtype: bool

Recall that we call a series that is all booleans a "mask". Only indexes with a "True" in the mask result in a row being included in the result.

In [232]:
iris[mask]

sepal length  sepal width  petal length  petal width           class
128           6.2          NaN           4.8          1.8  Iris-virginica

and thus we can see our problem.  There is a number missing in row 128.  Check the original dataframe to see for yourself.  We can also use `loc` or `iloc` to see the rows in this vicinity now we have identified the problem

In [233]:
iris.loc[120:130,:]

sepal length  sepal width  petal length  petal width           class
120           7.7          2.6           6.9          2.3  Iris-virginica
121           6.0          2.2           5.0          1.5  Iris-virginica
122           6.9          3.2           5.7          2.3  Iris-virginica
123           5.6          2.8           4.9          2.0  Iris-virginica
124           7.7          2.8           6.7          2.0             6.3
125           6.3          2.7           4.9          1.8  Iris-virginica
126           6.7          3.3           5.7          2.1  Iris-virginica
127           7.2          3.2           6.0          1.8  Iris-virginica
128           6.2          NaN           4.8          1.8  Iris-virginica
129           6.1          3.0           4.9          1.8  Iris-virginica
130           6.4          2.8           5.6          2.1  Iris-virginica

# Exercise - wide petals

Identify all rows where the "petal width" is greater than 5.  How many are there?

In [234]:
# put your answer in this code block
answer = iris[iris['petal width'] > 5.0]
print(answer)
print('')
print(str(answer['petal width'].count())+' row has a "petal width" greater than 5')

     sepal length  sepal width  petal length  petal width           class
134           6.4          2.8           5.6         22.0  Iris-virginica

1 row has a "petal width" greater than 5


## Fixing the error

What do to about the error is up to you.  You should create a "clean" data frame, separate to the other one regardless of your decision.  Possible choices are:
  * remove that whole row
  * remove that whole column
  * choose a value for the missing entry

We will demonstrate each.


In [235]:
# remove that row

# we need the opposite mask. A trick to do this is to perform the "equals false" operation :)
mask_valid = mask == False

# we can then use this as a mask to get only the rows we want.
clean_iris = iris[mask_valid]
clean_iris

sepal length  sepal width  petal length  petal width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
147           6.7          3.0           5.2          2.3  Iris-virginica
148           6.3          2.5           5.0          1.9  Iris-virginica
149           6.5          3.0           5.2          2.0  Iris-virginica
150           6.2          3.4           5.4          2.3  Iris-virginica
151           5.9          3.0           5.1          1.8  Iris-virginica

[151 rows x 5 columns]

In [236]:
# alternative!  Since we know exactly what row to drop, we can use the drop function

clean_iris = iris.drop(index=126)
clean_iris

sepal length  sepal width  petal length  petal width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
147           6.7          3.0           5.2          2.3  Iris-virginica
148           6.3          2.5           5.0          1.9  Iris-virginica
149           6.5          3.0           5.2          2.0  Iris-virginica
150           6.2          3.4           5.4          2.3  Iris-virginica
151           5.9          3.0           5.1          1.8  Iris-virginica

[151 rows x 5 columns]

In [237]:
# alternative - dropna does _all_ the hard work for us
clean_iris = iris.dropna()
clean_iris

sepal length  sepal width  petal length  petal width           class
0             5.1          3.5           1.4          0.2     Iris-setosa
1             4.9          3.0           1.4          0.2     Iris-setosa
2             4.7          3.2           1.3          0.2     Iris-setosa
3             4.6          3.1           1.5          0.2     Iris-setosa
4             5.0          3.6           1.4          0.2     Iris-setosa
..            ...          ...           ...          ...             ...
147           6.7          3.0           5.2          2.3  Iris-virginica
148           6.3          2.5           5.0          1.9  Iris-virginica
149           6.5          3.0           5.2          2.0  Iris-virginica
150           6.2          3.4           5.4          2.3  Iris-virginica
151           5.9          3.0           5.1          1.8  Iris-virginica

[151 rows x 5 columns]

In [238]:
# remove whole column with drop

clean_iris = iris.drop(columns="sepal width")
clean_iris

sepal length  petal length  petal width           class
0             5.1           1.4          0.2     Iris-setosa
1             4.9           1.4          0.2     Iris-setosa
2             4.7           1.3          0.2     Iris-setosa
3             4.6           1.5          0.2     Iris-setosa
4             5.0           1.4          0.2     Iris-setosa
..            ...           ...          ...             ...
147           6.7           5.2          2.3  Iris-virginica
148           6.3           5.0          1.9  Iris-virginica
149           6.5           5.2          2.0  Iris-virginica
150           6.2           5.4          2.3  Iris-virginica
151           5.9           5.1          1.8  Iris-virginica

[152 rows x 4 columns]

In [239]:
# choose a value for the missing entry with `fillna`
clean_iris = iris.fillna(0)
clean_iris.loc[125:128]

sepal length  sepal width  petal length  petal width           class
125           6.3          2.7           4.9          1.8  Iris-virginica
126           6.7          3.3           5.7          2.1  Iris-virginica
127           7.2          3.2           6.0          1.8  Iris-virginica
128           6.2          0.0           4.8          1.8  Iris-virginica

We used a couple of helper functions. [`dropna` will remove any rows with empty data](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html) while [`fillna` will replace any empty values with some other value](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html).

I tend not to fill in missing values, I tend to drop the whole row - that should be your default option but don't do it without paying attention to what is dropped and why.  Too many people just apply `dropna` without thinking.

# Exercise - find errors

Identify the erroneous data in the `class` column and remove that row.

**Advanced** Imagine you had not seen the error, what type of pandas code could you construct to find such an error.  I would suggest extracting that column as a series, get all the unique values in the series (`drop_duplicates` will help).

In [240]:
# row 124 has been identified with ```print(iris["class"].drop_duplicates())```; 
print(iris["class"].drop_duplicates())
print("")

# confirm row 124 is in fact erroneous
print(iris.iloc[124])
print("")

# drop erroneous row 124
iris = iris.drop(index=124)

# repeat identification process
print(iris["class"].drop_duplicates())
print("")

# now we can see there only three 'valid' classes for analysis


0          Iris-setosa
52     Iris-versicolor
102     Iris-virginica
124                6.3
Name: class, dtype: object

sepal length    7.7
sepal width     2.8
petal length    6.7
petal width     2.0
class           6.3
Name: 124, dtype: object

0          Iris-setosa
52     Iris-versicolor
102     Iris-virginica
Name: class, dtype: object



# Exercise - fishing

**Advanced** Find any other erroneous data and fix it.

In [241]:
# row 134 contains a 'petal width' of 22.0, which appears to be an unreasonable outlier value;  An uinformed opinion may be that, this value is missing it's decimal placement (ie. 22.0 may actual be 2.2)  
print(iris["petal width"].drop_duplicates())
print("")

# drop empty value row 134
iris = iris.drop(index=134)



# row 128 contains a 'sepal width' of NaN;  Suggestion: droprow
print(iris["sepal width"].drop_duplicates())
print("")

# drop empty value row 128
iris = iris.drop(index=128)



0       0.2
5       0.4
6       0.3
9       0.1
25      0.5
45      0.6
52      1.4
53      1.5
55      1.3
58      1.6
59      1.0
71      1.1
72      1.8
75      1.2
79      1.7
102     2.5
103     1.9
104     2.1
106     2.2
112     2.0
116     2.4
117     2.3
134    22.0
Name: petal width, dtype: float64

0      3.5
1      3.0
2      3.2
3      3.1
4      3.6
5      3.9
6      3.4
8      2.9
10     3.7
16     4.0
17     4.4
20     3.8
25     3.3
34     4.1
35     4.2
43     2.3
56     2.8
59     2.4
61     2.7
62     2.0
64     2.2
71     2.5
81     2.6
128    NaN
Name: sepal width, dtype: float64



# Exercise

There is a very useful method available on dataframes called "describe".  Below is an example of its use.

In [245]:

# let's assume lithgow is file "IDCJAC0009_063226_1800_Data.csv"
lithgow = pd.read_csv("data/rainfall/IDCJAC0009_063226_1800_Data.csv")
lithgow.describe()



Bureau of Meteorology station number          Year         Month  \
count                               53560.0  53560.000000  53560.000000   
mean                                63226.0   1950.823096      6.513667   
std                                     0.0     42.332504      3.447212   
min                                 63226.0   1878.000000      1.000000   
25%                                 63226.0   1914.000000      4.000000   
50%                                 63226.0   1951.000000      7.000000   
75%                                 63226.0   1987.000000     10.000000   
max                                 63226.0   2024.000000     12.000000   

                Day  Rainfall amount (millimetres)  \
count  53560.000000                   38268.000000   
mean      15.727502                       2.209170   
std        8.799543                       6.925849   
min        1.000000                       0.000000   
25%        8.000000                       0.000000   
50%       16.000000                       0.000000   
75%       23.000000                       0.800000   
max       31.000000                     244.100000   

       Period over which rainfall was measured (days)  
count                                    16050.000000  
mean                                         1.009470  
std                                          0.331585  
min                                          1.000000  
25%                                          1.000000  
50%                                          1.000000  
75%                                          1.000000  
max                                         19.000000

Experiment with this method on data you know well, [check the documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html).  How do you think this method can help you find erroneous data in your DataFrames?

# **I do think this method can help me find erroneous data in the DataFrame.**

## *Response:*

ATTRIBUTE                                          |   COMMENT
:---                                               |   :---:
Rainfall amount (millimetres)                      |   244.10 == max()
Rainfall amount (millimetres)                      |   15,292 == isnull()
Period over which rainfall was measured (days)     |   19.00 == max()
Period over which rainfall was measured (days)     |   37,510 == isnull()

# Concept Summary
  * We can index into a table with a mask
  * We can use boolean operators to create useful masks
  * `head`, `tail`, and `info` are useful for learning about your data
  * `dropna`, `drop_duplicates`, `duplicates`, `fillna` are usefull for cleaning your data

# Python concepts
  * `head`, `tail`, `info`, `dropna`, `drop_duplicates`, `duplicates`, `fillna` are all _methods_ on the data frame object
  * most of the methods we used returned entirely fresh values which we needed to capture in a new variable.  The original data frame was not changed by methods like `fillna`.
  * the `drop` method of a dataframe can take different parameters to do different things.  `index=` will drop a row, `columns=` will drop a column.  Both versions return a whole new DataFrame, leaving the original untouched.


# End of Material